### DISCUSSION TITLE
range search performance

### DISCUSSION THREAD

**Discussion by alkkkj** [hse-project/hse#40] [Q&A]
I noticed keys in CN tree were spilled in different child nodes by hash function,  right?  If so, how's the range search performance? Thanks.

**Comment by smoyerx** [comment]
Cursor iteration (range search) performance is excellent provided you use cursors with KVS storing multi-segment keys and with a filter whose length is equal to or greater than the key prefix length.  You can find detailed documentation with multiple examples in the [project documentation](https://hse-project.github.io/).  See the sections **Develop HSE Apps -> Concepts** and **Develop HSE Apps -> Best Practices**.

**Reply by alkkkj** [reply]
Do we have API for range query?  and  document for describing CN data structure?

**Reply by smoyerx** [reply]
Cursors are used for range queries.  There is an overview of cursor operation in the Concepts section of the project documentation, a detailed description of the cursor APIs in the [hse.h](https://github.com/hse-project/hse/blob/master/include/hse/hse.h) header file, and an example usage in the [samples](https://github.com/hse-project/hse/tree/master/samples) directory.

Regarding the cN data structure, we anticipate including more architecture documents in future releases, but for now you'll need to read the comments in the source code.
